cmd/kubeadm/app/util/etcd/etcd.go

```diff
@@ -124,6 +128,66 @@ func NewFromStaticPod(endpoints []string, manifestDir string, certificatesDir st
 	return New(endpoints, "", "", "")
 }
 
+// NewFromPods creates a GenericClient for the pods in the namespace kube-system and with label
+// component=etcd, manifestDir, and certificatesDir
+func NewFromPods(client clientset.Interface, manifestDir string, certificatesDir string) (*Client, error) {
+
+	// Gets the list of pods
+	glog.V(1).Infoln("getting the list of pods with component=etcd")
+	listOpts := metav1.ListOptions{LabelSelector: "component=etcd"}
+	pods, err := client.CoreV1().Pods(metav1.NamespaceSystem).List(listOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	// Note for reviewers: I'm not sure this is the best method for getting the endpoint
+	// out of etcd pods info... opinions?
+	endpoints := []string{}
+	for _, pod := range pods.Items {
+		endpoints = append(endpoints, fmt.Sprintf("https://%s:2379", pod.Status.PodIP))
+	}
+	glog.V(1).Infof("existing etcd endpoints read from pods: %s", strings.Join(endpoints, ","))
+
+	return New(
+		endpoints,
+		filepath.Join(certificatesDir, constants.EtcdCACertName),
+		filepath.Join(certificatesDir, constants.EtcdHealthcheckClientCertName),
+		filepath.Join(certificatesDir, constants.EtcdHealthcheckClientKeyName),
+	)
+}
+
+// AddMember notifies an existing etcd cluster that a new member is joining
+func (c Client) AddMember(name string, peerAddrs string) (map[string]string, error) {
+	cli, err := clientv3.New(clientv3.Config{
+		Endpoints:   c.Endpoints,
+		DialTimeout: 5 * time.Second,
```

Comment:
i wonder if the connection times increase with the number of etcd pods.

Reply:
I would up the timeout for new connection.  Default client inside the api-server is 20 seconds.

Reply:
increased timeout to 20